opentelemetry/src/main/java/io/grpc/opentelemetry/OpenTelemetryTracingModule.java

```diff
@@ -0,0 +1,408 @@
+                  tracerFactory.callEnded(status);
+                  super.onClose(status, trailers);
+                }
+              },
+              headers);
+        }
+      };
+    }
+  }
+
+  // Attribute named "message-size" always means the message size the application sees.
+  // If there was compression, additional event reports "message-size-compressed".
+  //
+  // An example trace with message compression:
+  //
+  // Sending:
+  // |-- Event 'Outbound message sent', attributes('sequence-numer' = 0, 'message-size' = 7854,
+  //                                               'message-size-compressed' = 5493) ----|
+  //
+  // Receiving:
+  // |-- Event 'Inbound compressed message', attributes('sequence-numer' = 0,
+  //                                                    'message-size-compressed' = 5493 ) ----|
+  // |-- Event 'Inbound message received', attributes('sequence-numer' = 0,
+  //                                                  'message-size' = 7854) ----|
+  //
+  // An example trace with no message compression:
+  //
+  // Sending:
+  // |-- Event 'Outbound message sent', attributes('sequence-numer' = 0, 'message-size' = 7854) ---|
+  //
+  // Receiving:
+  // |-- Event 'Inbound message received', attributes('sequence-numer' = 0,
+  //                                                  'message-size' = 7854) ----|
+  private void recordOutboundMessageSentEvent(Span span,
+      int seqNo, long optionalWireSize, long optionalUncompressedSize) {
+    AttributesBuilder attributesBuilder = io.opentelemetry.api.common.Attributes.builder();
+    attributesBuilder.put("sequence-number", seqNo);
+    if (optionalUncompressedSize != -1) {
+      attributesBuilder.put("message-size", optionalUncompressedSize);
+    }
+    if (optionalWireSize != -1 && optionalWireSize != optionalUncompressedSize) {
+      attributesBuilder.put("message-size-compressed", optionalWireSize);
+    }
+    span.addEvent("Outbound message sent", attributesBuilder.build());
+  }
+
+  private void recordInboundCompressedMessage(Span span, int seqNo, long optionalWireSize) {
+    AttributesBuilder attributesBuilder = io.opentelemetry.api.common.Attributes.builder();
+    attributesBuilder.put("sequence-number", seqNo);
+    attributesBuilder.put("message-size-compressed", optionalWireSize);
+    span.addEvent("Inbound compressed message", attributesBuilder.build());
+  }
+
+  private void recordInboundMessageSize(Span span, int seqNo, long bytes) {
+    AttributesBuilder attributesBuilder = io.opentelemetry.api.common.Attributes.builder();
+    attributesBuilder.put("sequence-number", seqNo);
+    attributesBuilder.put("message-size", bytes);
+    span.addEvent("Inbound message received", attributesBuilder.build());
+  }
+
+  private String generateErrorStatusDescription(io.grpc.Status status) {
+    if (status.getDescription() != null) {
+      return status.getCode() + ". " + status.getDescription();
```

Comment:
nit: We generally use `:` between the code and description.

TIL Java now lets you do string concatenation with a non-string left operand.